[WIP] Integration with Datafusion

marvinlanhenke opened the pull request:
I had some fun, working with Datafusion and I want to share the (rough) design with you.

As outlined in #242 I went ahead and:

- created a new crate `integrations`
- a subfolder `datafusion` with its respective modules (catalog.rs, schema.rs, etc.)
- implemented some of the traits e.g. `CatalogProvider`
- basic test infra for integration test with HiveMetastore

---

The overall structure should be extensible, meaning - if we want to support an integration with e.g. `polars` we simply create a new subfolder and implement the traits. 

The test infra is supposed to be used by every integration. 

The IcebergCatalogProvider, for example, should be usable by any Catalog (since it Arc<dyn Catalog> for the client (dynamic dispatch))

I really would appreciate your thoughts on this approach - so we can discuss how to proceed, whats missing upstream (like #277), etc.

marvinlanhenke commented:
@liurenjie1024 @ZENOTME @Fokko 
PTAL and let me know what you think.

liurenjie1024 commented:
Thanks @marvinlanhenke This is amazing, I'll take a review later!

tshauck commented on Cargo.toml:
```diff
@@ -21,6 +21,7 @@ members = [
     "crates/catalog/*",
     "crates/examples",
     "crates/iceberg",
+    "crates/integrations",
```
Perhaps as a V2 of this integration it'd make sense to use features and/or move the datafusion dep down into the integrations crate (or maybe have a crate per integration).

To your point about polars or other extensions, if this concept grew the dependency load would get challenging for folks who don't need it.

marvinlanhenke replied:
yes - I think this makes a lot of sense. I'd definitely go for this design. thanks

marvinlanhenke replied:
@tshauck 
I updated the structure - to allow a specific crate per integration;
I'm not sure if I handled the feature-flags correctly - so PTAL and let me know what you think.

tshauck replied:
Updating the integrations to be datafusion specific looks good to me. For the actual datafusion dependency, you could move that into the Cargo.toml of the new datafusion subcrate itself. So e.g...

```toml
[dependencies]
dashmap = { workspace = true }
datafusion = "37.0.0"
futures = { workspace = true }
iceberg = { workspace = true }
log = { workspace = true }
tokio = { workspace = true }
```

I think it's generally considered better to scope subcrate specific dependencies to the crate that needs it. I'm sorry for muddying the waters via the feature suggestion, but then I'm not sure the additional complexity of it would be needed.

marvinlanhenke replied:
> I'm sorry for muddying the waters via the feature suggestion, but then I'm not sure the additional complexity of it would be needed.

Perhaps @liurenjie1024 @Xuanwo you can take another look - whats best practice here in terms of deps?

tshauck commented on crates/integrations/src/datafusion/schema.rs:
```diff
@@ -0,0 +1,97 @@
+    }
+}
+
+impl SchemaProvider for IcebergSchemaProvider {
```
I realize these are just stubs, I'd be happy to help fill in the guts here and for the `TableProvider` traits.

marvinlanhenke replied:
yes - I just wanted to share a draft as early as possible - in order to get the design ideas aligned. And sure - any help is appreciated since I'm not so familiar with the Datafusion traits and I think we also need to implement `ExecutionPlan` in order to provide the correct return type for the `TableProvider`...?

tshauck replied:
Sweet, happy to help... for the catalog data [delta-rs](https://github.com/delta-io/delta-rs/blob/5eade5e1f07e80ecd053bd538e005894b7e88a8f/crates/core/src/data_catalog/unity/datafusion.rs#L140) is probably a good example to crib from. Also, using `async_trait` will improve the aesthetics of `IcebergSchemaProvider::table` and `IcebergTableProvider::scan`.

For execution then (`ExecutionPlan`), to your point it needs to be implemented. Similarly [`DeltaTable`](https://github.com/delta-io/delta-rs/blob/3094bd28cee94b41b502cff6a479965d1e81e990/crates/core/src/delta_datafusion/mod.rs#L654-L705) implements `TableProvider` which uses [`DeltaScan`](https://github.com/delta-io/delta-rs/blob/3094bd28cee94b41b502cff6a479965d1e81e990/crates/core/src/delta_datafusion/mod.rs#L789-L798) that implements `ExecutionPlan`.

What makes sense to you help-wise, so I don't step on your toes? Maybe I could branch off this and take a crack at `IcebergScan` or just commit to this branch or something else?

marvinlanhenke replied:
you can simply pick what you like - and let me know, how you'd want to contribute - if you want to commit to this branch, I guess I need to invite you to my fork?

tshauck replied:
Ok great, I'll get started on an `IcebergScan` implementation, or at least plan out some work in more detail and follow up if there's something ambiguous.

tshauck replied:
I looked a little more at this, and it seems reasonable to get a first version up and running. Blocker right now is https://github.com/apache/iceberg-rust/pull/277 like you mentioned.

tshauck reviewed (commented):
This is really cool! I've been hoping for I wouldn't have to try something like this 😅... I left a comment about package structure imagining if I were to use this.

Also if you'all do want to move forward with this, I'd be happy to help fill-in some of the datafusion traits implementations.

simonvandel commented on crates/integrations/datafusion/src/catalog.rs:
```diff
@@ -0,0 +1,67 @@
+        for name in schema_names {
+            let provider =
+                IcebergSchemaProvider::try_new(client.clone(), &NamespaceIdent::new(name.clone()))
+                    .await?;
```
Drive-by review:
Since `IcebergSchemaProvider::try_new` is async, I'm guessing it might take some time to execute?

In that case, instead of awaiting in a for-loop, perhaps using https://docs.rs/futures/latest/futures/future/fn.join_all.html or similar would make sense?

ZENOTME commented:
Thanks! Sorry for replying late. I think this is a good start for the integration work. And I have completed https://github.com/apache/iceberg-rust/pull/277. Maybe we can convert this PR to ready for review now. @marvinlanhenke

marvinlanhenke commented:
> Thanks! Sorry for replying late. I think this is a good start for the integration work. And I have completed #277. Maybe we can convert this PR to ready for review now. @marvinlanhenke

thanks for the feedback - i made some minor changes based on the suggestions; and converted this PR into ready for review. If we agree on the basic design here, we could merge and split the actual impl into multiple issues.

@liurenjie1024 @ZENOTME @Fokko @Xuanwo @sdd @tshauck PTAL - especially, the part about "feature-flags" I have no idea whats best practice in 🦀

liurenjie1024 commented on crates/integrations/datafusion/Cargo.toml:
```diff
@@ -0,0 +1,46 @@
+# under the License.
+
+[package]
+name = "iceberg-integrations-datafusion"
```
```suggestion
name = "iceberg-datafusion"
```
How about just this?

liurenjie1024 commented on crates/integrations/datafusion/Cargo.toml:
```diff
@@ -0,0 +1,46 @@
+
+[features]
+default = []
+datafusion = ["dep:datafusion"]
```
Sorry I'm a little confusing here, as a crate for integration with datafusion, why this is a feature rather enabled by default?

marvinlanhenke replied:
Here I'm probably even more confused; I tried this based on the review by @tshauck - but I have no experience with this and I'm not sure how to structure the crate, in order to allow a possible end-user to download only the integration dependency he needs e.g. iceberg-datafusion or iceberg-polars. etc.

So here I'm hoping for feedback and help.

Xuanwo replied:
I believe we should have `integrations/datafusion` and `integrations/polars`. So this crate `iceberg-integrations-datafusion` should enable `datafusion` by default, and doesn't need switch.

tshauck replied:
Sorry for muddying the waters. I tried to suggest not going the feature route based [here](https://github.com/apache/iceberg-rust/pull/324#discussion_r1562767139) (near the end). I was a little confused by this project having all dependencies in the workspace's Cargo.toml even if only used by one crate.

marvinlanhenke replied:
> I believe we should have `integrations/datafusion` and `integrations/polars`. So this crate `iceberg-integrations-datafusion` should enable `datafusion` by default, and doesn't need switch.

@Xuanwo 
thank you - then I guess my last commit fixed it.

liurenjie1024 commented on crates/integrations/datafusion/README.md:
```diff
@@ -0,0 +1,22 @@
+
+# Apache Iceberg Integrations
+
+This crate contains the official Native Rust implementation of Apache Iceberg Integrations.
```
```suggestion
This crate contains integration of apache datafusion and apache iceberg.
```

viirya replied:
Apache DataFusion and Apache Iceberg

liurenjie1024 commented on crates/integrations/datafusion/src/table.rs:
```diff
@@ -0,0 +1,79 @@
+        todo!()
+    }
+
+    fn scan<'life0, 'life1, 'life2, 'life3, 'async_trait>(
```
We should remove these auto generate lifetimes

liurenjie1024 commented on crates/integrations/datafusion/src/table.rs:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    fn schema(&self) -> datafusion::arrow::datatypes::SchemaRef {
+        todo!()
```
I think the convertion between iceberg schema and arrow schema has been implemented?

marvinlanhenke replied:
yes; i believe we merged this 2 days ago; so i can take a look at this again now

marvinlanhenke replied:
done

liurenjie1024 commented on crates/integrations/datafusion/src/table.rs:
```diff
@@ -0,0 +1,79 @@
+        'life3: 'async_trait,
+        Self: 'async_trait,
+    {
+        todo!()
```
I think this could be implemented with current scan api without filter pushdown?

marvinlanhenke replied:
...I think this has to be done (PTAL @liurenjie1024):

- impl asyn fn scan() on IcebergTableProvider which has to return an `Arc<dyn Execution Plan>`
  - therefore we need to: impl the trait ExecutionPlan for IcebergTableScan which is basically a wrapper around an actual PhysicalPlan e.g. ParquetExec 
    - therefore we need to: create an PhysicalPlan when building our IcebergTableScan

I think creating the PhysicalPlan and providing all the input like FileScanConfig will be the majority of the work and I'm not sure we can provide all of this yet...but well see...

reference from delta-rs:
- [DeltaScan](https://github.com/delta-io/delta-rs/blob/main/crates/core/src/delta_datafusion/mod.rs#L790-L798)
- [ExecutionPlan for DeltaScan](https://github.com/delta-io/delta-rs/blob/main/crates/core/src/delta_datafusion/mod.rs#L814-L881)
- [building DeltaScan](https://github.com/delta-io/delta-rs/blob/main/crates/core/src/delta_datafusion/mod.rs#L507-L651) and the actual PhysicalPlan // [ParquetExec](https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/parquet/struct.ParquetExec.html)

liurenjie1024 replied:
I took a look at the `ExecutionPlan` trait, and I think the main part if the `execute` method which could return an arrow record batch stream, which has been implemented by `TableScan::to_arrow`. Other parts, such as partitioning, ordering we can leave it as simple as possible. 
I agree that this is not blocker and we can move it to next pr.

marvinlanhenke replied:
I don't think this will work?

```rust
fn execute(...) -> Result<SendableRecordBatchStream> {
  ...
  // I can't await here since fn execute() needs to remain sync
  // using tokio::Runtime block_on() seems weird?
  table_scan.to_arrow().await.unwrap()
  // also the return type is BoxStream<'static, Result<RecordBatch>>
  // which does not satisfy SendableRecordBatchStream since we need to
  // impl RecordBatchStream as well
}
```

...so, how to avoid those issues? 

We could try to impl RecordBatchStream on TableScan itself, this however would require us to have datafusion as a dependency in the core and we'd still have the async issue - so I don't think this would be a good solution. 

Or we try the same approach as [delta-rs](https://github.com/delta-io/delta-rs/blob/main/crates/core/src/delta_datafusion/mod.rs#L627-L642), and create an actual datafusion physical plan and wrap this:

```rust
struct IcebergTableScan {
  plan: Arc<dyn ExecutionPlan>
  ...
}

impl ExecutionPlan for IcebergTableScan {
  ...
  fn execute(...) -> Result<SendableRecordBatchStream> {
     self.plan.execute()  
  }
}
```

... what do you think @liurenjie1024 ?

simonvandel replied:
I think you can do what is documented here: https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#lazily-async-create-a-stream

So inside `IcebergTableProvider::scan` you create `IcebergScanExec` where `IcebergScanExec` implements `ExecutionPlan`.

The `ExecutionPlan::execute` impl of `IcebergScanExec` would be something like this

```rust
impl ExecutionPlan for IcebergScanExec {
  fn execute(
        &self,
        partition: usize,
        context: Arc<TaskContext>
    ) -> Result<SendableRecordBatchStream> {
    let fut = create_batch_stream(self.table_scan);

    let stream = futures::stream::once(fut).try_flatten();
    Ok(Box::pin(RecordBatchStreamAdapter::new(self.schema.clone(), stream)))
  }
}
```

/// async function that returns a stream
async fn create_batch_stream(table: TableScan) -> Result<SendableRecordBatchStream> {
    todo!()
}


(Not @liurenjie1024 but hopefully it is helpful - I didn't compile it :) )

marvinlanhenke replied:
@simonvandel thanks a lot - this looks promising

liurenjie1024 replied:
Thanks @simonvandel for the comments! Yeah, that looks great to me.

liurenjie1024 commented on crates/integrations/datafusion/src/schema.rs:
```diff
@@ -0,0 +1,102 @@
+        self.tables.get(name).is_some()
+    }
+
+    fn table<'life0, 'life1, 'async_trait>(
```
We should remove these auto generated lifetimes to make it more readable.

liurenjie1024 commented on crates/integrations/datafusion/src/table.rs:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    fn table_type(&self) -> datafusion::datasource::TableType {
+        todo!()
```
I think `TableType::Base` is the answer.

liurenjie1024 commented on crates/integrations/datafusion/src/catalog.rs:
```diff
@@ -0,0 +1,73 @@
+            .list_namespaces(None)
+            .await?
+            .iter()
+            .flat_map(|ns| ns.as_ref().clone())
```
I'm not sure if datafusion allows nested namespace here, if not, maybe we should check namespace name.

marvinlanhenke replied:
In this case; we provide the catalog (e.g. hive metastore) and we don't support nested namespaces. 

Even if datafusion allows nested namespaces; once we hit our catalog provider we'd only return an empty list. 
So I don't think we need to perform any checks here?

liurenjie1024 replied:
Sounds reasonable to me.

liurenjie1024 reviewed (commented):
Thanks @marvinlanhenke for this pr! This is really excited for me since this will provide the foundation of sql interface against iceberg-rust! It's really amazing! I've left some comment to improve, but it looks great! I'll invite datafusion community to help review.

marvinlanhenke commented:
> I've left some comment to improve, but it looks great! I'll invite datafusion community to help review.

@liurenjie1024 
Thanks for the review.
I fixed most of the basic issues regarding structure, naming, async-trait etc.

I'll work on the missing impls over the next couple of days - and see how far we can get with the current state of iceberg-rust.

simonvandel commented on crates/integrations/datafusion/src/catalog.rs:
```diff
@@ -0,0 +1,95 @@
+    /// A concurrent `HashMap` where keys are namespace names
+    /// and values are dynamic references to objects implementing the
+    /// [`SchemaProvider`] trait.
+    schemas: DashMap<String, Arc<dyn SchemaProvider>>,
```
Would a std Hashmap suffice? I don't see the concurrency properties of DashMap being used.

I might be missing something though.

marvinlanhenke replied:
the datafusion docs provide a link to a reference implementation by [delta-rs](https://github.com/delta-io/delta-rs/blob/951436ecec476ce65b5ed3b58b50fb0846ca7b91/crates/deltalake-core/src/data_catalog/unity/datafusion.rs#L21), that uses a DashMap as well. 

My understanding based on the [docs](https://docs.rs/datafusion/latest/datafusion/catalog/trait.CatalogProvider.html#implementing-remote-catalogs) is that datafusion might share the `CatalogProvider` on multiple threads inside an Arc; thus concurrently accessing our "cached" schemas, etc.? Perhaps someone else can clarify this?

> [...] The pattern that DataFusion itself uses to plan SQL queries is to walk over the query to [find all schema / table references in an async function](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionState.html#method.resolve_table_references), performing required remote catalog in parallel, and then plans the query using that snapshot.

simonvandel replied:
Will the map ever be mutated?
It might be, I was just not seeing it in this PR.

If not, then a std Hashmap seems sufficient to me.

simonvandel replied:
Cool, well then I think a std Hashmap should be fine, since no actual concurrent operations need to happen

liurenjie1024 commented on crates/integrations/datafusion/src/catalog.rs:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    fn schema_names(&self) -> Vec<String> {
+        self.schemas.keys().cloned().collect()
```
I'm thinking that this maybe incorrect since others processes may create new namespaces after creating the `try_new` method? We should the result of `list_namespaces` each time. For performance issue, we may create sth like [`CachingCatalog` in java](https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/core/src/main/java/org/apache/iceberg/CachingCatalog.java#L46) to implement `Catalog` trait, what do you think?

marvinlanhenke replied:
yes, you're right this is a simple but naive impl. based on the [docs ](https://docs.rs/datafusion/latest/datafusion/catalog/trait.CatalogProvider.html#implementing-remote-catalogs) and the ref from delta-rs. 

I think having a caching wrapper would be a better solution. 

However, since we have to provide a vec of all schema_names we cannot check if a single schema is in the cache and if not fetch again. So I'm guessing we can only implement it with a time based eviction policy i.e. cache all schemas for 1min in order to avoid multiple network calls in a short amount of time (same concept as debouncing)?

liurenjie1024 replied:
Yes, when adding a cache, we need to a design for the trade of between performance and consistency. But I think we could leave it to actualy cache implentation? For this pr, we should not cache them in the providers, e.g. `CatalogProvider`, `SchemaProvider`, etc, but calling `Catalog` api directly.

marvinlanhenke replied:
sure, we can do it like that for this PR and impl. the more sophisticated solution later.

marvinlanhenke replied:
@liurenjie1024 
...I took another look; and I'm not sure we can call the Catalog API directly each time the funtion on the e.g. `trait CatalogProvider` is invoked. The trait is required to be synchronous; however our catalog api calls are async.

So perhaps; we leave it as is for this PR and work on the cache instead of wasting time and effort on another intermediate solution? But perhaps there is an obvious solution to this I'm missing?

liurenjie1024 replied:
Yes, you are right. As of now, we can only provide a snapshot version. It's possible to block on an async call, but it requires runtime api. cc @Xuanwo @Fokko @viirya What dou you think? My suggestion is to leave it as it now, and fix it when we have runtime api landed.

marvinlanhenke replied:
> when we have runtime api landed

is there any issue or discussion already related to that? Or would you mind to quickly outline whats the plan here?

Xuanwo replied:
> It's possible to block on an async call, but it requires runtime api.

It does possible but we should avoid using `rt.block_on()` in our lib. 

> My suggestion is to leave it as it now, and fix it when we have runtime api landed.

I'm second with this suggestion. We can move on and figure how to improve this part in the future.

Fokko replied:
We should be hesitant with caching, especially avoiding upfront optimizations. For Iceberg in general, consistency is king.

> For this pr, we should not cache them in the providers, e.g. CatalogProvider, SchemaProvider, etc, but calling Catalog api directly.

This makes sense, but I see the issue with blocking on async calls. At first, I would take the price of waiting for the blocking calls. Even it is still a remote call, the ones to the REST catalog should be lightning-fast (that's where the caching happens).

viirya commented on crates/integrations/datafusion/src/schema.rs:
```diff
@@ -0,0 +1,95 @@
+        self
+    }
+
+    fn table_names(&self) -> Vec<String> {
+        self.tables.keys().cloned().collect()
+    }
+
+    fn table_exist(&self, name: &str) -> bool {
+        self.tables.get(name).is_some()
+    }
```
These implementations might have similar issue? The table names and table existence is not up to date when calling.

marvinlanhenke commented:
@liurenjie1024 @ZENOTME @viirya @simonvandel 
...I just went ahead and pushed the recent updates; PTAL

## unresolved / todo:
- [ ] proper integration tests for table scan (requires us to setup a table with actual snapshots / data )
- [ ] a cache for `IcebergCatalogProvider` and `IcebergSchemaProvider` / so data does not become stale
- [ ] improve impl ExecutionPlan for IcebergTableScan (once we support filter pushdown, etc.)

marvinlanhenke commented on crates/integrations/datafusion/src/physical_plan/scan.rs:
```diff
@@ -0,0 +1,123 @@
+        .to_arrow()
+        .await
+        .map_err(to_datafusion_error)?
+        .map_err(to_datafusion_error);
```
@liurenjie1024 @ZENOTME @viirya 
I had no idea how to solve this better; but it still looks weird and redundant.

Why do I need to do this? First `map_err` unwraps the result from `to_arrow()` and the second `map_err` converts the outer result (wrapping the boxed stream) into a `datafusion::error:Result` which is required by `futures::stream::once(fut).try_flatten()`.

liurenjie1024 replied:
I think it's ok for now. There is [flatten](https://doc.rust-lang.org/std/result/enum.Result.html#method.flatten) api to simplify it, but it's not stable. We can refine it in future.

liurenjie1024 commented:
> @liurenjie1024 @ZENOTME @viirya @simonvandel ...I just went ahead and pushed the recent updates; PTAL
> 
> ## unresolved / todo:
> * [ ]  proper integration tests for table scan (requires us to setup a table with actual snapshots / data )
> * [ ]  a cache for `IcebergCatalogProvider` and `IcebergSchemaProvider` / so data does not become stale
> * [ ]  improve impl ExecutionPlan for IcebergTableScan (once we support filter pushdown, etc.)

Hi, @marvinlanhenke How about opening a tracking issue to track the issues related to datafusion integration? I guess we will have more things to do, and we can edit on that tracking issue.

liurenjie1024 commented on crates/integrations/datafusion/src/physical_plan/scan.rs:
```diff
@@ -0,0 +1,123 @@
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        Ok(self)
```
I think this should never be called, how about throwing an error here?

marvinlanhenke replied:
I thought the same; then I checked `CsvExec` in Datafusion since it's also a leaf-node that should also not have any children. turns out they do the same and return `Ok(self)` as well. [link](https://github.com/apache/datafusion/blob/main/datafusion/core/src/datasource/physical_plan/csv.rs#L181-L186)

liurenjie1024 reviewed (approved):
Thanks @marvinlanhenke for this pr!

liurenjie1024 commented:
Let's wait a moment to see what others think about the catalog snapshot problem. cc @Xuanwo @Fokko @viirya @sdd PTAL

marvinlanhenke commented:
> Let's wait a moment to see what others think about the catalog snapshot problem. cc @Xuanwo @Fokko @viirya @sdd PTAL

If it turns out, that the PR is okay (for now) - let me update the missing docs; before we merge - should make the follow up work a lot easier in a couple of days/weeks.

Xuanwo commented on crates/integrations/datafusion/src/schema.rs:
```diff
@@ -0,0 +1,99 @@
+            .collect();
+
+        Ok(IcebergSchemaProvider {
+            tables: tables.into_iter().collect(),
```
Seems we can collect into `HashMap` directly without into `Vec<_>` first?

marvinlanhenke replied:
yes, this was a left-over from before where I used DashMap. thanks for spotting that.

Xuanwo commented on crates/integrations/datafusion/src/schema.rs:
```diff
@@ -0,0 +1,99 @@
+        self.tables.get(name).is_some()
+    }
+
+    async fn table(&self, name: &str) -> datafusion::error::Result<Option<Arc<dyn TableProvider>>> {
```
We can `use datafusion::error::Result as DfResult`, what do you think?

marvinlanhenke replied:
yes - i like that. fixed.

Xuanwo reviewed (commented):
Thanks for raising this PR. Mostly LGTM, just some minor nits. 

I'm looking forward to work with you to integrate FileIO into it.

Fokko commented on crates/integrations/datafusion/src/catalog.rs:
```diff
@@ -0,0 +1,97 @@
+    /// collects these providers into a `HashMap`.
+    pub async fn try_new(client: Arc<dyn Catalog>) -> Result<Self> {
+        // TODO:
+        // Schemas and providers should be cached and evicted based on time
+        // As of right now; schemas might become stale.
```
> _There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors._ - Leon Bambrick

Is there no way to leave this up to the user?

marvinlanhenke replied:
> _There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors._ - Leon Bambrick

😄the classic.

I think leaving it up to the user, leads us to the issue about blocking an async call in a sync trait function? I think if we have an idea how to handle this, we can better reason about if, when, and where to cache?

from the docs:
> To implement [CatalogProvider](https://docs.rs/datafusion/latest/datafusion/catalog/trait.CatalogProvider.html) and [SchemaProvider](https://docs.rs/datafusion/latest/datafusion/catalog/schema/trait.SchemaProvider.html) for remote catalogs, you need to provide an in memory snapshot of the required metadata. Most systems typically either already have this information cached locally or can batch access to the remote catalog to retrieve multiple schemas and tables in a single network call.

Fokko commented on crates/integrations/datafusion/src/physical_plan/scan.rs:
```diff
@@ -0,0 +1,136 @@
+
+use crate::to_datafusion_error;
+
+#[derive(Debug)]
```
Can't we enforce this kind of thing in a linter? :)

marvinlanhenke replied:
> Seem this feature is still under discussion: [rust-lang/rustfmt#3744](https://github.com/rust-lang/rustfmt/issues/3744)

uff with no update since 2020?
I rearranged the stmt to go with the suggested approach.

Fokko reviewed (approved):
LGTM, thanks a bunch working on this @marvinlanhenke. I don't have much to comment here since it is more on the Datafusion side, rather than the Iceberg side 👍

marvinlanhenke commented:
I've addressed those issues raised by @Xuanwo - thanks for the review.

I think we can merge this PR for now - and continue working on #357; 
especially the caching issues; and proper integration testing with data.

@Fokko @liurenjie1024 
I haven't checked now, but from memory I believe py-iceberg uses pyspark to setup tables with actual data for proper integration testing? Would this make sense to tackle such an issue in the next weeks in order to provide proper integration testing? Or should we wait once #349 lands and we have a dedicated crate for e2e testing?

viirya commented:
Thanks for working on this @marvinlanhenke . Looks good to me. There are some more integrations with DataFusion I'm interested to work with like improving ExecutionPlan for IcebergTableScan. Let's move this forward.
